crates/beautifier/src/lib.rs

```diff
@@ -0,0 +1,292 @@
+/*
+ * Copyright 2022 Fluence Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#![deny(
+    dead_code,
+    nonstandard_style,
+    unused_imports,
+    unused_mut,
+    unused_variables,
+    unused_unsafe,
+    unreachable_patterns
+)]
+
+use std::fmt::Display;
+use std::io::{Error as IoError, Result as IoResult, Write};
+
+use air_parser::ast::{
+    Ap, Call, Fail, FoldScalar, FoldStream, Instruction, Match, MisMatch, New, Next, Null, Par,
+    Seq, Triplet, Value, Xor,
+};
+
+pub const DEFAULT_INDENT_SIZE: usize = 4;
+
+macro_rules! multiline {
+    ($beautifier:expr, $indent:expr, $fmt1:literal $(, $arg1:expr)*; $nest1:expr) => ({
+        let out = &mut $beautifier.output;
+        $crate::fmt_indent(out, $indent)?;
+        writeln!(out, $fmt1 $(, $arg1)*)?;
+        $crate::Beautifier::beautify_walker($beautifier, $nest1, $indent + $beautifier.indent_size)
+    });
+    ($beautifier:expr, $indent:expr, $fmt1:literal $(, $arg1:expr)*; $nest1:expr; $fmt2:literal $(, $arg2:expr)*; $nest2:expr) => ({
+        let step = $beautifier.indent_size;
+        {
+            let out = &mut $beautifier.output;
+            $crate::fmt_indent(&mut *out, $indent)?;
+            writeln!(out, $fmt1 $(, $arg1)*)?;
+        }
+        crate::Beautifier::beautify_walker(&mut *$beautifier, $nest1, $indent + step)?;
+        {
+            let out = &mut $beautifier.output;
+            $crate::fmt_indent(out, $indent)?;
+            writeln!(out, $fmt2 $(, $arg2)*)?;
+        }
+        crate::Beautifier::beautify_walker(&mut *$beautifier, $nest2, $indent + step)
+    });
+}
+
+fn fmt_indent(output: &mut impl Write, indent: usize) -> IoResult<()> {
```

Comment:
You could use the following approach for indentation:
```rust
write!(w, "{:indent$}Indented text!", "", indent=indent)
```

[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=ab951cb07db3b1415f048341eeda42f7)